### ISSUE TITLE
vplan review for RV32 interrupts for cv32e40x

### ISSUE THREAD

**Issue by strichmo** [openhwgroup/core-v-verif#884]
### Task Outcome
Updated vplan coverage plan definition for interrupts for the CV32E40X.

### Background information
This review will update the coverage definition for the CV32E40X interrupt implementation.  

A description of the known changes can be found in this draft PR:
https://github.com/openhwgroup/core-v-verif/pull/880

### Location Information
The updated vplan Excel spreadsheet can be fetched from this branch:

https://github.com/silabs-robin/core-v-verif/blob/irq_vplan/cv32e40x/docs/VerifPlans/Simulation/interrupts/CV32E40X_interrupts.xlsx

The final vplan will be updated into the cv32e40x/dev branch (to be propagated to master via scheduled merge cycles).

### Completion Criteria
The accepted review criteria are completed and the final reviewed vplan is placed in the above directory.

**Comment by silabs-robin** [comment]
> Only mtvec_addr_i[31:8] bits are used (maybe we should mandate that the other bits are required to be 0 by the environment?)

If it is not illegal then I think it should be tested.
Should it be illegal? Why? What could happen upon violation?


> Why are all branches (BEQ, BNE, etc.) treated separately?

It was like that from before, when the vplan was copied from 40P.
The important point seems to be that both taken/nontaken is tested.
I would prefer to change it, make it more terse, if there is no good reason to keep it.


> Does 'all fence instructions' include fence.i?

I think there are more problems with these wordings...
Anyway, I propose that it can be just "All RV32I instructions interrupted" and "All RV32I branch instructions interrupted both taken and nontaken".
And also add "RV32Zifencei", "X interface", "illegal". Then similarly update "RV32C" and the others.

> What are 'counter instructions'

This item should be removed, in my opinion. Unless there is some reason to keep it?
Does it mean to interrupt csr instructions using counters csrs specifically?
It was likely included in the 40P vplan because Counters are listed among the supported extensions.

> Illegal instructions must be interruptible [...]
> Instructions offloaded via the eXtension interface must be interruptible [...]
> We need some kind of progress assertion [...]

I'll add them.

**Comment by Silabs-ArjanB** [comment]
> 
>     Only mtvec_addr_i[31:8] bits are used (maybe we should mandate that the other bits are required to be 0 by the environment?)
> 
> If it is not illegal then I think it should be tested.
> Should it be illegal? Why? What could happen upon violation?

The other bits are simply ignored.

> It was like that from before, when the vplan was copied from 40P.
> The important point seems to be that both taken/nontaken is tested.
> I would prefer to change it, make it more terse, if there is no good reason to keep it.

There is no reason to change it, but it seems rather arbitrary. Agreed that the taken - non-taken is important though.

>     What are 'counter instructions'
> 
> This item should be removed, in my opinion. Unless there is some reason to keep it?
> Does it mean to interrupt csr instructions using counters csrs specifically?
> It was likely included in the 40P vplan because Counters are listed among the supported extensions.
> 

I have no idea what 'counter instructions' is supposed to mean or why they were considered special. The 'counters' chapter from the unpriv spec is an extension indeed (that does not have a proper name, but we sometimes call it Zicount). That is maybe why it was treated separately. If so, then it is a good idea indeed to check that for all extensions that they can be interrupted.

**Comment by silabs-robin** [comment]
> It would be good to check that a fence.i that is in the shadow of a taken branch or jump (or interrupt or debug entry) is fully killed and so does not raise the request signal (this could also be in the **interrupt vplan** as similar checks are needed for all possible side effects of instructions that should not occur if an instruction is supposed to be killed).

For the interrupt vplan, do you mean that _interrupts_ kill the instruction (instead of _branching_ killing the instruction) and then no side-effects were visible?
I am not sure what specifically can be done about this. There are already these "all X instructions get interrupted" which intends to check this. And there are the assertions (in the vplan) for each feature, like how the fence.i assertions on req/ack will demand that a req eventually leads to fence.i-branch taken (so if a killing has a req side-effect then there will be loud assertions problems).
Otherwise, would one have to comb through each instruction and see which side-effects to explicitly include in the interrupts vplan?
It could be a good idea to do what you suggest but I struggle to visualize what it could entail.

**Comment by Silabs-ArjanB** [comment]
>>For the interrupt vplan, do you mean that interrupts kill the instruction (instead of branching killing the instruction) and then no side-effects were visible?

In all cases (interrupts, exceptions, branches, etc.) the _controller_ kills the instructions.

Somewhere we need to check that killed instructions do not have side effects. Maybe the interrupt vplan is not the right place to do this, I agree with that. My point is that it is easy to check that for example a load causes a data_req_o = 1; the things that should also be checked is that data_req_o = 0 if no load is executed (e.g. because it is killed due to branch, interrrupt, debug entry, etc.)

**Comment by strichmo** [comment]
Thanks again @silabs-robin 

- For NMIs, mepc, mcause and mstatus (mpie and mie) should all be checked in the ISS
- Let's drop the "counter" instructions line.  That's a leftover that was effectively folded into Zicsr instructions in the E40P
-- Should we think about adding another instruction coverage specific to CSR writes crossed with valid CSRs.  In other words, ensure that the instruction both "before" and "after (or interrupted" by an IRQ or WFI are CSR writes to every possible valid CSR (mstatus, mie, mcounteren, etc.)  We should have all of the data structures and sampling for this already.
- I am not exactly sure what is intended by "Is this OK to move this item up..."  My intention is that individual lines of a good vplan should map to as simple a check, coverage, etc. as possible.  
- Add an assertion check that nmip can be asserted in either debug mode or non-debug mode (coverage).

**Comment by strichmo** [comment]
Regarding @Silabs-ArjanB comment above on killed instructions:
This is the appropriate vplan for this kind of coverage.  

We should absolutely include the "interrupt latency due to instruction" check Arjan mentions in his initial comment.

**Comment by silabs-robin** [comment]
@strichmo 

> For NMIs, mepc, mcause and mstatus (mpie and mie) should all be checked in the ISS

Fixed.


> Let's drop the "counter" instructions line

Fixed.


> Should we think about adding another instruction coverage specific to CSR writes crossed with valid CSRs

Could be a good idea. How great is the value of doing this? How easy/costly would it be to do?


> I am not exactly sure what is intended by "Is this OK to move this item up..."

Old comment. I cleaned it up now.


> Add an assertion check that nmip can be asserted in either debug mode or non-debug mode (coverage)

Fixed.


> on killed instructions: This is the appropriate vplan for this kind of coverage

This is difficult.
One could say "Interrupted instructions get killed and have no side-effects (e.g. data_req_o == 0)", but that is too general and not very actionable. When one checks not only correctness of "somesignal == 1" but "somesignal == 0" it begins to look like equivalency checking vs a model, in my mind. That's why I struggle to find an appropriate scope for this sort of checking.
I can think more about it on monday, but if anyone has suggestions for how to phrase this kind of coverage then that is much appreciated.


> include the "interrupt latency due to instruction" check Arjan mentions in his initial comment

If this is the _"We need some kind of progress assertion [...]"_ then I have added items for it now.
However, I phrased it as "within a fixed number of cycles" as I don't know appropriate numbers for it.
